File: scripts/start-terminals.js
const { exec } = require('child_process');
const path = require('path');
const { fileURLToPath } = require('url');

// 需要修改 __dirname 的获取方式，因为在 CommonJS 中它是全局变量
// const __dirname = path.dirname(fileURLToPath(import.meta.url));
const rootDir = path.resolve(__dirname, '..');

const commands = [
  { cmd: 'yarn watch', cwd: rootDir },
  { cmd: 'yarn watch', cwd: path.join(rootDir, 'extensions/xunlongaicode') },
  { cmd: 'yarn build && yarn dev', cwd: path.join(rootDir, 'extensions/xunlongaicode/webview-ui') },
  { cmd: 'yarn watch', cwd: path.join(rootDir, 'extensions/orangepicode-core') },
  { cmd: 'yarn build && yarn dev', cwd: path.join(rootDir, 'extensions/orangepicode-core/webview-ui') }
];

// 添加延迟函数
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

// 使用异步函数按顺序启动终端
async function startTerminals() {
  for (let i = 0; i < commands.length; i++) {
    const { cmd, cwd } = commands[i];
    const escapedCmd = cmd.replace(/"/g, '\\"');
    const escapedCwd = cwd.replace(/"/g, '\\"');

    const appleScript = `
      tell application "Terminal"
        do script "cd \\"${escapedCwd}\\" && ${escapedCmd}"
      end tell
    `;

    try {
      await new Promise((resolve, reject) => {
        exec(`osascript -e '${appleScript}'`, (error, stdout, stderr) => {
          if (error) {
            console.error(`启动终端 ${i} 失败: ${error.message}`);
            reject(error);
            return;
          }
          console.log(`已在新终端 ${i} 中启动: ${cmd}`);
          resolve();
        });
      });

      // 每次启动终端后等待一段时间
      await delay(1000);
    } catch (err) {
      console.error(`终端 ${i} 启动过程中出错: ${err}`);
    }
  }
}

startTerminals().catch(err => {
  console.error('启动终端过程中发生错误:', err);
});

File: scripts/start_terminals.py
import os
import subprocess
import time
import platform
import sys

# 获取根目录路径
root_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# 定义要执行的命令
commands = [
    {"cmd": "yarn watch", "cwd": root_dir},
    {"cmd": "yarn watch", "cwd": os.path.join(root_dir, 'extensions/xunlongaicode')},
    {"cmd": "yarn build && yarn dev", "cwd": os.path.join(root_dir, 'extensions/xunlongaicode/webview-ui')},
    {"cmd": "yarn watch", "cwd": os.path.join(root_dir, 'extensions/orangepicode-core')},
    {"cmd": "yarn build && yarn dev", "cwd": os.path.join(root_dir, 'extensions/orangepicode-core/webview-ui')}
]

def start_terminal_macos(cmd, cwd):
    """在macOS上启动新终端并执行命令"""
    escaped_cmd = cmd.replace('"', '\\"')
    escaped_cwd = cwd.replace('"', '\\"')

    apple_script = f'''
    tell application "Terminal"
      do script "cd \\"{escaped_cwd}\\" && {escaped_cmd}"
    end tell
    '''

    try:
        subprocess.run(['osascript', '-e', apple_script], check=True)
        print(f"已在新终端中启动: {cmd}")
        return True
    except subprocess.CalledProcessError as e:
        print(f"启动终端失败: {e}")
        return False

def start_terminal_windows(cmd, cwd):
    """在Windows上启动新终端并执行命令"""
    try:
        subprocess.Popen(f'start cmd.exe /K "cd /d "{cwd}" && {cmd}"',
                         shell=True, cwd=cwd)
        print(f"已在新终端中启动: {cmd}")
        return True
    except Exception as e:
        print(f"启动终端失败: {e}")
        return False

def start_terminal_linux(cmd, cwd):
    """在Linux上启动新终端并执行命令"""
    try:
        # 尝试使用gnome-terminal
        subprocess.Popen(['gnome-terminal', '--', 'bash', '-c', f'cd "{cwd}" && {cmd}; exec bash'])
        print(f"已在新终端中启动: {cmd}")
        return True
    except FileNotFoundError:
        try:
            # 尝试使用xterm
            subprocess.Popen(['xterm', '-e', f'cd "{cwd}" && {cmd}; bash'])
            print(f"已在新终端中启动: {cmd}")
            return True
        except FileNotFoundError:
            print("无法找到合适的终端模拟器")
            return False

def main():
    """主函数，根据操作系统启动终端"""
    system = platform.system()

    for i, command in enumerate(commands):
        cmd = command["cmd"]
        cwd = command["cwd"]

        print(f"正在启动终端 {i+1}/{len(commands)}...")

        if system == "Darwin":  # macOS
            success = start_terminal_macos(cmd, cwd)
        elif system == "Windows":
            success = start_terminal_windows(cmd, cwd)
        elif system == "Linux":
            success = start_terminal_linux(cmd, cwd)
        else:
            print(f"不支持的操作系统: {system}")
            sys.exit(1)

        if not success:
            print(f"终端 {i+1} 启动失败，继续下一个...")

        # 每次启动终端后等待一段时间
        time.sleep(1)

if __name__ == "__main__":
    main()
